### ISSUE TITLE
XML support with libxml(2)

### ISSUE THREAD

**Issue by emacstheviking** [part-cw/lambdanative#78]
I started to plan a port of an RSS reader I wrote.... no xml!

I have spent time looking at the library / module structure and it doesn't look like it should be *too hard* to add libxml (http://www.xmlsoft.org/) into the mix... but I think it would take me ages to fully understand the magical package download and build stuff and then add Scheme wrappers to the calls to just decode an RSS or general XML document.

So... as a nice to have not too far in the future, what are the odds? In the meantime, I shall attempt to do it myself and see how I get on!

Thanks again,
Sean Charles.

**Comment by mgorges** [comment]
@emacstheviking have you considered [modules/ssax](blob/master/modules/ssax/ssax.scm) for this purpose? I previously used it to parse the response from an arterial blood pressure monitor. It'll transform XML into SXML. It takes LONG to compile but might work for your purpose. If this won't work for you we can have a look at libxml2, which has an acceptable license, but I am unfamiliar with its interface so writing a wrapper might take a while.  

Finally, I don't know why the `ssax` module isn't documented in the Wiki but will coordinate with @clpetersen to get that resolved. Sorry, Matthias

**Comment by emacstheviking** [comment]
ssax ---> Ummm, that's handy... I didn't see that... I think that's me
happy then!

Thanks

**Comment by emacstheviking** [comment]
Here it is!

https://github.com/emacstheviking/lambdanative-repl

It is *very rough* and probably a little flaky but my current knowledge of
the Scheme R5RS standard is raw, as is my experience with Gambit scheme so
I am pleased I got this far with it.

It leverages 99.9% of the standard cmuscheme.el code, all I did was study
hard that code and the code in comint.el and realise that all I needed was
a tiny bit of elisp to fire off a connection to the backend REPL code.

I wasted many hours of my life studying "geiser" code, "jlongtser" code
from his iPhone app and many other sites before deciding it would be just
easier to shut it all out at roll my own.

The backend REPL code is a single file, it started out as the sample
"tcprepl" code but I have hacked, slashed and burned it into something, er,
different!

I have no doubt you guys there will weep, cry and cringe when you see the
code but it's early days for me.

But... IT WORKS!

I have added code to a running console application and then executed it
from the application REPL and vice versa: adding a simple function in the
application and calling into from the comint *scheme* buffer.

I tries to be graceful but there a lots of edge cases that can end up in
having to "kill" the buffer and launch it again.

BUT I can edit code now, something I have been waiting to do.. as I edit my
application I will also maintain / grow this as well.

All the best,
Sean Charles.

**Comment by clpetersen** [comment]
Hi Sean,

Awesome work! I'm currently on the road, but will be putting this to the test as soon as I get back next week. Perhaps this could even become an integral part of developing and debugging apps in lambdanative.. I'll have a look at how we might roll it into the main system. Thanks for all your effort!

Chris

**Comment by mgorges** [comment]
@emacstheviking sorry for the delayed response,  I tried it yesterday but being unfamiliar with Emacs couldn't get it to work for me - something about the scheme executable not being found. - I'll sit down with @clpetersen on Monday to take a look. But the fact that it works for you is quite exciting!

**Comment by emacstheviking** [comment]
Matthias,

It *works* for me... however this weekend a failed El Capitan upgrade has
left me totally machineless. I am furious that Apple have released it and
it has bricked my machine, I am typing thi son recovery mode.

As for the Emacs hting... don't get too excited! It's rough, bleeding edge
and still requires a lot of work. What puzzles me is that people already
use Emacs with scheme and so I don't understand why just running the bog
standard Gambit REPL code against comit doesn't work.

I have days and days and days trying to get other peoples existing code to
work but ultimately ended up being both frustrated and disapointed. I rool
my own code only when I have to. Comint mode is already used by the
cmuscheme mode and after studying the code I realised it all should work

BUT... there is work to do on the application side as I didn't realise
until after I had posted that actually, some of the scheme commands from
Emacs wrap up the outbound code in a (compile ) term and so C-x C-e doesn't
work, the only thing that works is to "send region to REPL" which is what I
think I managed to get working.

Once I get a working mAc back I shall continue my work, goodjob it is on
github because I think I am going to have to wipe the entire hard rive and
start again. something nasty has happened and again, I am very angry at QA
from Apple

Sean

**Comment by mgorges** [comment]
Sean, sorry about your computer troubles. We made it a firm decision to not upgrade until later in the game, when the initial issues are solved (so like 10.11.3/4) - in fact I just got Yosemite last week to look at the posted bug and others in the team still use Mavericks. It is hard to balance the gain of some cool new features with the risk of bricking your development tools. 

As for the REPL I'll meet with Chris next week and have a look at your Emacs integration piece. In the meantime enjoy the rest of your Sunday

**Comment by clpetersen** [comment]
To follow up on the original topic, I've added an libxml2 based module `xml` as an alternative to `ssax` in commit 954b189. I'll close this XML issue, and we can continue discussing the REPL in a dedicated issue as needed.

**Comment by emacstheviking** [comment]
Fantastic. You guys certainly are committed to lambdanative that's for sure.

I am still off the air, almost back with a working iMac... then I will also
resume my work on the Emacs//REPL although I'd like to make it more
portable than just Emacs. Ultimately it shouldn't care!

All the best again,
Sean

**Comment by emacstheviking** [comment]
Hello,

You guys have done a fantastic job on the REPL... it's way better than
anything I could (and did) produce on my own. I have my iMac back working
again and tonight I have been training myself on using the scheme REPL you
have given to us.

I have it all working against DemoRedSquare but there is a potential issue
somewhere... when I submit a command using C-x C-e from a Scheme file, the
REPL buffer windows takes anything up to five seconds or so before a
response comes back, almost as if some time out is in progress.

If I run the same command using telnet from the command line, there is no
delay and everything is instant response as you would expect.

The code in ln_repl.scm appears to contain no timeouts I can see, I'm
guessing you've started a `comint` buffer mode around a telnet session,
something I never considered at all. Nice idea that.

The delay is very frustrating though, I am trying to learn the saxx module
(which is still not on the documentation list by the way, nor is libxml2)
and having to wait five seconds every time I issue a REPL command is
slowing it all down. Feels like the 80-s again!

So, have I done something wrong or is there an issue?

All the best and once again congratulations on more sterling work!

Sean Charles.

**Comment by clpetersen** [comment]
Hi Sean

Great that you are trying this out! All I did was roll together a bunch of existing emacs modes. I was hoping that an actual emacs user, like yourself, would be able to hack it into shape, perhaps using the new comint communication concepts that you started working on before. Anyhow, I'll try to replicate this delay issue.

Both `ssax` and `xml` (the name of the module that uses libxml2) are documented on the wiki. However, the full api is not described there, only the exposed/commonly used interface. You are very welcome to expand on this. The API is fully described on the respective official homepages of the libraries.

Chris

**Comment by clpetersen** [comment]
I just played around with the REPL in DemoRedSquare.. I don't have a delay when submitting code. I was initially wondering if this could be a thread preempting issue on the gambit side, but since you have no delay in a telnet session, it sounds more like an emacs issue.. I've seen reports that comint is slow when dealing with long strings... do you see the delay even when sending small snippets, like `(+ 1 2)`?

**Comment by emacstheviking** [comment]
Chris,

Today I tried again and all was fine... I was trying to work out how to use
the ssax module and I pulled back a big RSS feed from the BBC news site.
Maybe there was a connection. Since then I have not seen that problem
manifest itself again so maybe it is all good and well after all. Then I
realised that (sxml:document "ur') did it all and I was blown away!

As for the "comint" thing... I believe the cmuscheme.el mode you have built
upon already makes use of it... all I did was figure out a specific way of
getting it to talk to my junk.

As a died-in-the-wool Emacs user, using what you have feels very natural
and comfortable to me at least.

I am now attempting a "Serious" project; a simple RSS reader that uses the
glgui-list thing. I am also going to try to document my proceedings,
potentially using "noweb" as I have used Literate Programming techniques
many times in the past for small self contained projects that require both
code and detailed documentation.

I am sure I will be in touch at some point, I am off into hack space
again....  :)

Many thanks
Sean
